lib/client/session.go

```diff
@@ -593,41 +621,57 @@ func (ns *NodeSession) pipeInOut(shell io.ReadWriteCloser) {
 			log.Errorf(err.Error())
 		}
 	}()
-	// copy from the local input to the remote shell:
-	go func() {
-		defer ns.closer.Close()
-		buf := make([]byte, 128)
-
-		stdin := ns.terminal.Stdin()
-		if ns.terminal.IsAttached() && ns.enableEscapeSequences {
-			stdin = escape.NewReader(stdin, ns.terminal.Stderr(), func(err error) {
-				switch err {
-				case escape.ErrDisconnect:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\n%v\r\n", err)
-				case escape.ErrTooMuchBufferedData:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\nremote peer may be unreachable, check your connectivity\r\n", trace.Wrap(err))
-				default:
-					fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\n", trace.Wrap(err))
+
+	switch mode {
+	case types.SessionObserverMode, types.SessionModeratorMode:
+		go func() {
+			defer ns.closer.Close()
+
+			handleNonPeerControls(mode, ns.terminal, func() {
+				_, err := sess.SendRequest(teleport.ForceTerminateRequest, true, nil)
+				if err != nil {
+					fmt.Printf("\n\rError while sending force termination request: %v\n\r", err.Error())
 				}
-				ns.closer.Close()
 			})
-		}
-		for {
-			n, err := stdin.Read(buf)
-			if err != nil {
-				fmt.Fprintf(ns.terminal.Stderr(), "\r\n%v\r\n", trace.Wrap(err))
-				return
+		}()
+	case types.SessionPeerMode:
+		// copy from the local input to the remote shell:
+		go func() {
+			defer ns.closer.Close()
+			buf := make([]byte, 128)
+
+			stdin := ns.terminal.Stdin()
+			if ns.terminal.IsAttached() && ns.enableEscapeSequences {
+				stdin = escape.NewReader(stdin, ns.terminal.Stderr(), func(err error) {
+					switch err {
+					case escape.ErrDisconnect:
+						fmt.Fprintf(ns.terminal.Stderr(), "\r\n%v\r\n", err)
+					case escape.ErrTooMuchBufferedData:
+						fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\nremote peer may be unreachable, check your connectivity\r\n", trace.Wrap(err))
+					default:
+						fmt.Fprintf(ns.terminal.Stderr(), "\r\nerror: %v\r\n", trace.Wrap(err))
+					}
+					ns.closer.Close()
+				})
 			}
 
-			if n > 0 {
-				_, err = shell.Write(buf[:n])
+			for {
+				n, err := stdin.Read(buf)
```

Comment:
The docs for `io.Reader` say:

> Callers should always process the n > 0 bytes returned before considering
    the error err. Doing so correctly handles I/O errors that happen after
    reading some bytes and also both of the allowed EOF behaviors.

Should we handle the data first and then check the error?